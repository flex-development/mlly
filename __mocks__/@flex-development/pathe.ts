/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
