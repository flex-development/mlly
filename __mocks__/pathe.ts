/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
