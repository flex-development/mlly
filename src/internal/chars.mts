/**
 * @file Internal - chars
 * @module mlly/internal/chars
 */

/**
 * Characters that may have special meaning.
 *
 * @enum {string}
 */
enum chars {
  asterisk = '*',
  at = '@',
  empty = '',
  hash = '#'
}

export default chars
