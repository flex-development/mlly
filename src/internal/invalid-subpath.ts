/**
 * @file Internal - invalidSubpath
 * @module mlly/internal/invalidSubpath
 */

import chars from '#internal/chars'
import {
  ERR_INVALID_MODULE_SPECIFIER,
  type ErrInvalidModuleSpecifier
} from '@flex-development/errnode'
import type { ModuleId } from '@flex-development/mlly'
import { fileURLToPath } from '@flex-development/pathe'
import { ok } from 'devlop'

/**
 * Create an {@linkcode ERR_INVALID_MODULE_SPECIFIER} error for an invalid
 * subpath pattern match.
 *
 * @see {@linkcode ErrInvalidModuleSpecifier}
 * @see {@linkcode ModuleId}
 *
 * @internal
 *
 * @param {string} subpath
 *  Subpath pattern
 * @param {string} patternMatch
 *  Subpath pattern match
 * @param {ModuleId} packageUrl
 *  URL of directory containing `package.json` file
 * @param {boolean | null | undefined} [isImports]
 *  Whether `target` is internal to the package
 * @param {ModuleId | null | undefined} [parent]
 *  URL of parent module
 * @return {ErrInvalidModuleSpecifier}
 *  New node error
 */
function invalidSubpath(
  subpath: string,
  patternMatch: string,
  packageUrl: ModuleId,
  isImports?: boolean | null | undefined,
  parent?: ModuleId | null | undefined
): ErrInvalidModuleSpecifier {
  ok(patternMatch, 'expected `patternMatch`')

  return new ERR_INVALID_MODULE_SPECIFIER(
    subpath.replace(chars.asterisk, () => patternMatch),
    `request is not a valid match in pattern "${subpath}" for the ` +
      `"${isImports ? 'imports' : 'exports'}" resolution of ` +
      `${fileURLToPath(new URL('package.json', packageUrl))}`,
    parent && fileURLToPath(parent)
  )
}

export default invalidSubpath
