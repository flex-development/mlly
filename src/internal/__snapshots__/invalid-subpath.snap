// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/invalidSubpath > should return new \`ERR_INVALID_MODULE_SPECIFIER\` (0) 1`] = `Invalid module './node_modules/esbuild/lib/main.js' request is not a valid match in pattern "./*" for the "exports" resolution of \${process.cwd()}/__fixtures__/node_modules/build-exports/package.json`;

exports[`unit:internal/invalidSubpath > should return new \`ERR_INVALID_MODULE_SPECIFIER\` (1) 1`] = `Invalid module '#fixtures/node_modules/subpath-imports/package.json' request is not a valid match in pattern "#fixtures/*" for the "imports" resolution of \${process.cwd()}/package.json`;
